src/datasets/arrow_dataset.py

```diff
@@ -3029,6 +3032,135 @@ def select(
         if len(self) == 0:
             return self
 
+        # If the indices are contiguous, simply slice the arrow table
+        if isinstance(indices, range):
+            if _is_range_contiguous(indices) and indices.start >= 0:
+                start, length = indices.start, indices.stop - indices.start
+                return self._select_contiguous(start, length, new_fingerprint=new_fingerprint)
+        else:
+            try:
+                start = next(iter(indices))
+            except StopIteration:
+                return self._select_contiguous(0, 0, new_fingerprint=new_fingerprint)
+            if start >= 0:
+                counter_from_start = itertools.count(start=start)
+                if all(i == j for i, j in zip(indices, counter_from_start)):
```

Comment:
Also note this implementation has an overhead for `np.array` and `pd.Series`, compared to regular Python `list`:
```python
In [34]: def check_with_counter(indices):
    ...:     start = next(iter(indices))
    ...:     counter_from_start = itertools.count(start=start)
    ...:     if all(i == j for i, j in zip(indices, counter_from_start)):
    ...:         return True
    ...:     else:
    ...:         return False

In [81]: lis = list(range(10_000_000))

In [82]: %timeit check_with_counter(lis)
657 ms ± 5.34 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [83]: arr = np.array(range(10_000_000))

In [84]: %timeit check_with_counter(arr)
2.43 s ± 37.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [85]: ser = pd.Series(list(range(10_000_000)))

In [86]: %timeit check_with_counter(ser)
1.35 s ± 18.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Thanks for the info ! In the docstring of `select` we should maybe encourage users to pass a `range` instead of a list/array/series in such cases anyway